src/Parsing/Impl/Definitions/IScopeNode.cs

```diff
@@ -0,0 +1,48 @@
+using Microsoft.Python.Parsing.Ast;
+
+namespace Microsoft.Python.Parsing {
+    /// <summary>
+    /// Represents a <see cref="Node"/> with a Scope
+    /// </summary>
+    public interface IScopeNode : INode {
+        /// <summary>
+        /// Scope node name
+        /// </summary>
+        string Name { get; }
+
+        /// <summary>
+        /// Parent of the current scope
+        /// </summary>
+        IScopeNode ParentNode { get; set; }
+
+        /// <summary>
+        /// Body of the scope node, null if not applicable 
+        /// </summary>
+        Statement Body { get; }
+
+        /// <summary>
+        /// Gloabl scope
+        /// </summary>
+        PythonAst GlobalParent { get; }
+
+        /// <summary>
+        /// Holds information about the scope gathered from Parser, mainly used in  <see cref="PythonNameBinder"/>
+        /// </summary>
+        ScopeInfo ScopeInfo { get; }
```

Comment:
Why `ScopeInfo` is a separate class? I.e. why not expose members here directly since every scope has `ScopeInfo` ?

Reply:
Basic rules on interfaces:
- on public there should be no setters unless there is a valid scenario for the consumer outside of the analysis engine. We provide evaluator on analysis that can be used for basic operations such as declaring variables.
- on public interface assume there is concurrency. So if setter is exposed, it must be thread-safe.
- on any interface enumeration must be safe. I.e. public interface user must be able to safely enumerate collections without worrying if they can be modified by the user code directly or indirectly via API calls or by concurrency.
- Above mostly applies to internal interfaces as well - otherwise why have interfaces...

Reply:
For the first, I added the `ScopeInfo` class so that different ScopeNode implementations can have the same ScopeInfo types. For example comprehension and functions both handle scopes the same ways, so instead of having both implement the same methods and have the same code body, they can just share `FunctionScopeInfo`

Reply:
Also I'll try to get to your comments by EOD tomorrow, thanks for the comments!